### ISSUE TITLE
glossary term syntax and automatic generation of glossary

### ISSUE THREAD

**Issue by jxxcarlson** [asciidoctor/asciidoctor#1488]
Implement a syntax like {{potrezebie}} which causes "potrezebie" to be added to an automatically generated glossary which appears at the end of a section.  The syntax 
{{potrezebie, fanciful contraption}} results in the entry

potrezebie: fanciful contraption
- Glossary terms are rendered in the text in red if the attribute :glossary-term-color:  is set.
- The glossary is constructed on a per section basis if :glossary: is set.
- See [this example](http://epsilon.my.noteshare.io/section/writing-tools?part=Glossary#_glossary) for a test implementation in Asciidoctor-LaTeX.

Additionally: items in the glossary should be links back to the marked terms in the text.  This 
is not yet implemented in  Asciidoctor-LaTeX.

**Comment by mojavelinux** [comment]
Whether or not we decide to have a shorthand syntax, we should definitely have a formal syntax.

```
glossterm:potrezebie[fanciful contraption]
```

Don't forget that it's possible to introduce either the formal or shorthand syntax today using an extension.

Another possible idea is to implement two-way linking between an explicit glossary and glossary terms. It really depends on where we think the canonical source of the content should be. Obviously, it would be easier to find terms in the source if they were in one place. But it's much more convenient for the author to introduce terms inline.

**Comment by jxxcarlson** [comment]
I’ll change my “gloss” to “glossterm” — at the moment I have
only one way linking — from glossary entry to term in the text.
Will follow your suggestion and implement the reverse as well.
There is an example at

	http://examples.noteshare.io/section/dante-inferno-canto-1# <http://examples.noteshare.io/section/dante-inferno-canto-1#>

that cries out for this kind of link.

**Comment by jxxcarlson** [comment]
```
module Asciidoctor::LaTeX
  class MacroPreprocessor < Asciidoctor::Extensions::Preprocessor


    def process document, reader
      regex = /{{(.*?)}}/
      return reader if reader.eof?
      replacement_lines = reader.read_lines.map do |line|
        if line.include? '{{'
          scan = line.scan regex
          scan.each do |match|
            target = match[0]
            puts "target: #{target}".red
            line = line.gsub("{{#{target}}}", "glossterm::[#{target}]")
          end
        end
        line
      end
      reader.unshift_lines replacement_lines
      reader
    end

  end
end

**Comment by jxxcarlson** [comment]
I have the two chunks of code below to manage glossary entries.  If in (2) I replace `glossterm::` by
`glossterm:`, then Asciidoctor-LaTeX leaves `{{potrezebie, fantastical machine}}` as
`glossterm:[potrezebie, fantastical machine`]  instead processing it into HTML.  What am I missing here?


```
# (1)
  class GlossInlineMacro <  Asciidoctor::Extensions::InlineMacroProcessor
    use_dsl
    named :glossterm
    def process parent, target, attributes
      term = attributes.values[0]
      id = term.gsub(' ', '_').gsub(/\W/, '')
      "<span class='glossary_term' id=gloss_#{id} >#{term}</span>"
    end
  end
```


```
# (2)
module Asciidoctor::LaTeX
  class MacroPreprocessor < Asciidoctor::Extensions::Preprocessor
    def process document, reader
      regex = /{{(.*?)}}/
      return reader if reader.eof?
      replacement_lines = reader.read_lines.map do |line|
        if line.include? '{{'
          scan = line.scan regex
          scan.each do |match|
            target = match[0]
            puts "target: #{target}".red
            line = line.gsub("{{#{target}}}", "glossterm::[#{target}]")
          end
        end
        line
      end
      reader.unshift_lines replacement_lines
      reader
    end

  end
end
```

**Comment by jxxcarlson** [comment]
Bidirectional links between glossary terms and their entries in the glossary now working at 
[http://examples.noteshare.io/section/dante-inferno-canto-1](Writing tool examples)

**Comment by remocrevo** [comment]
I'd also like to see this feature added to AsciiDoctor.  Any progress?

**Comment by luvarqpp** [comment]
Is there any progress? I would like to have some two way links and other similar things like in TeX.

Side-note, asciidoc manual seems have some info how to make glossary "by hand": https://asciidoctor.org/docs/user-manual/#user-glossary

**Comment by djencks** [comment]
I implemented something like this in javascript: https://gitlab.com/djencks/asciidoctor-glossary.  Mine works (to some extent) across pages in Antora.  It also shows the definition as a tooltip.  I'm not sure how to implement links back to the usages if there are more than one.

**Comment by mojavelinux** [comment]
> Side-note, asciidoc manual seems have some info how to make glossary "by hand": asciidoctor.org/docs/user-manual/#user-glossary

Correct. The glossary has always been a "by hand" thing in AsciiDoc. We're exploring the possibility of adding something that is auto-generated from the flow of content. But the idea is still incubating.

**Comment by djencks** [comment]
There's a question on back-links from the glossary, if there are multiple uses then what?  Dan suggested on Gitter that the first one be picked.  On a single page this makes some sense but that may not be well-defined in an multi-page environment such as Antora. A couple of possible solutions would be
- a 'primary' flag in glossterm to indicate this is where the back reference should go.
- popping up a list of back links with context, such as section headings they occur in.

**Comment by jubax** [comment]
I'm new to asciidoctor and just wanted to ask if there has been progress with the auto-generation of the glossary.

**Comment by mojavelinux** [comment]
The best place to ask questions is the project chat at https://chat.asciidoctor.org. If there is an update on an issue, there will be activity on it. If there is no activity, you can assume there is no progress. However, you can ask about it in the chat.

**Comment by DanielFran** [comment]
@mojavelinux I found this https://gitlab.com/djencks/asciidoctor-glossary from @djencks, but not sure it is still working with current version,
I try to use it within latest antora version but it did not work for me.
But this might be the MVP for this request?

**Comment by mojavelinux** [comment]
No, that's not the approach we are going to take.
